HeavyFlavorAnalysis/RecoDecay/interface/BPHDecayMomentum.h

```diff
@@ -0,0 +1,184 @@
+  /// add a previously reconstructed particle giving it a name
+  virtual void add( const std::string& name,
+                    const BPHRecoConstCandPtr& comp );
+
+  /// get a composite by the simple sum of simple particles
+  virtual const pat::CompositeCandidate& composite() const;
+
+  /// get the list of names of simple particles directly produced in the decay
+  /// e.g. in JPsi -> mu+mu-   returns the two names used for muons
+  ///      in B+   -> JPsi K+  returns the name used for the K+
+  ///      in Bs   -> JPsi Phi returns an empty list
+  virtual const std::vector<std::string>& daugNames() const;
+
+  /// get the list of names of previously reconstructed particles
+  /// e.g. in JPsi -> mu+mu-   returns an empty list
+  ///      in B+   -> JPsi K+  returns the name used for the JPsi
+  ///      in Bs   -> JPsi Phi returns the two names used for JPsi and Phi
+  virtual const std::vector<std::string>& compNames() const;
+
+  /// get the list of simple particles directly produced in the decay
+  /// e.g. in JPsi -> mu+mu-   returns the two muons
+  ///      in B+   -> JPsi K+  returns the K+
+  ///      in Bs   -> JPsi Phi returns an empty list
+  /// (clones are actually returned)
+  virtual const std::vector<const reco::Candidate*>& daughters() const;
+
+  /// get the full list of simple particles produced in the decay,
+  /// directly or through cascade decays
+  /// e.g. in B+   -> JPsi K+ ; JPsi -> mu+mu- returns the mu+, mu-, K+
+  ///      in Bs   -> JPsi Phi; JPsi -> mu+mu-; Phi -> K+K-
+  ///                                          returns the mu+, mu-, K+, K-
+  /// (clones are actually returned)
+  virtual const std::vector<const reco::Candidate*>& daughFull() const;
+
+  /// get the original particle from the clone
+  virtual const reco::Candidate* originalReco(
+          const reco::Candidate* daug ) const;
+
+  /// get the list of previously reconstructed particles
+  /// e.g. in JPsi -> mu+mu-   returns an empty list
+  ///      in B+   -> JPsi K+  returns the JPsi
+  ///      in Bs   -> JPsi Phi returns the JPsi and Phi
+  virtual const std::vector<BPHRecoConstCandPtr>& daughComp() const;
+
+  /// get a simple particle from the name
+  /// return null pointer if not found
+  virtual const reco::Candidate* getDaug( const std::string& name ) const;
+
+  /// get a previously reconstructed particle from the name
+  /// return null pointer if not found
+  virtual BPHRecoConstCandPtr getComp( const std::string& name ) const;
+
+ protected:
+
+  struct Component {
+    const reco::Candidate* cand;
+    double mass;
+    double msig;
+    std::string searchList;
+  };
+
+  // get an object filled in the constructor
+  // to be used in the creation of other bases of BPHRecoCandidate
+  const std::vector<Component>& componentList() const;
+
+  // map linking cloned particles to original ones
+  std::map<const reco::Candidate*, const reco::Candidate*> clonesMap;
```

Comment:
which one is the clone? the key or the value?
random ordering (based on memory pointer value) doesn't seem optimal and can also lead to non-reproducibilities.
If there is a predefined "native" index (like an index in the collection), it may be a better choice.